lightningd/peer_htlcs.c

```diff
@@ -637,7 +650,68 @@ static struct htlc_accepted_hook_response *
 htlc_accepted_hook_deserialize(const tal_t *ctx, const char *buffer,
 			       const jsmntok_t *toks)
 {
-	return NULL;
+	struct htlc_accepted_hook_response *response;
+	const jsmntok_t *resulttok, *failcodetok, *paykeytok, *chanupdtok;
+
+	if (!toks || !buffer)
+		return NULL;
+
+	resulttok = json_get_member(buffer, toks, "result");
+
+	/* If the result is "continue" we can just return NULL since
+	 * this is the default behavior for this hook anyway */
+	if (!resulttok) {
+		fatal("Plugin return value does not contain 'result' key %s",
+		      json_tok_full(buffer, toks));
+	}
+
+	if (json_tok_streq(buffer, resulttok, "continue")) {
+		return NULL;
+	}
+
+	response = tal(ctx, struct htlc_accepted_hook_response);
+	if (json_tok_streq(buffer, resulttok, "fail")) {
+		response->result = htlc_accepted_fail;
+		failcodetok = json_get_member(buffer, toks, "failure_code");
+		chanupdtok = json_get_member(buffer, toks, "channel_update");
+		if (!failcodetok || !json_to_number(buffer, failcodetok,
+						    &response->failure_code)) {
+			/* Use unknown payment hash since it has no
+			 * long lasting effect but will still prevent
+			 * the sender from retrying. In the newest
+			 * spec it is called "incorrect or unknown
+			 * payment details" which matches our use-case
+			 * pretty closely. */
+			response->failure_code =
+			    WIRE_INCORRECT_OR_UNKNOWN_PAYMENT_DETAILS;
```

Comment:
1.  It's OK to provide a default, but we should get upset if it's invalid, not silently replace it.
2.  An intermediary node MUST NOT return this; we should probably just use TEMPORARY_NODE_FAILURE which is always a valid response.